src/odbc_connection.cpp

```diff
@@ -2844,108 +2913,399 @@ fetch_and_store
+            // Instead, call SQLGetData, and adjust buffer size accordingly
+            if (data->columns[column_index]->is_long_data) {
+
+              // Get the first chunk of data
+              data->sqlReturnCode =
+              SQLGetData
+              (
+                data->hSTMT,
+                column_index + 1,
+                data->columns[column_index]->bind_type,
+                data->long_data_buffer,
+                data->long_data_buffer_size,
+                &data->long_data_string_length_or_indicator
+              );
+              printf("SQLGetData result: %d\n", data->sqlReturnCode);
+              if (!SQL_SUCCEEDED(data->sqlReturnCode))
+              {
+                return data->sqlReturnCode;
+              } 
+
+              // If the data is null, simply indicate and continue to the next
+              // column
+              if (data->long_data_string_length_or_indicator == SQL_NULL_DATA)
+              {
+                row[column_index].size = SQL_NULL_DATA;
+                continue;
+              } else if (data->long_data_string_length_or_indicator == SQL_NO_TOTAL) 
+              {
+                SQLLEN data_to_copy_length = 0;
+                while (data->long_data_string_length_or_indicator == SQL_NO_TOTAL)
                 {
-                  row[column_index].size = data->bound_columns[column_index].length_or_indicator_array[row_index];
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size]();
-                  memcpy(
-                    row[column_index].char_data,
-                    (SQLCHAR *)data->bound_columns[column_index].buffer + row_index * data->columns[column_index]->buffer_size,
-                    row[column_index].size
+                  switch(data->columns[column_index]->bind_type)
+                  {
+                    case SQL_C_BINARY:
+                      data_to_copy_length = data->long_data_buffer_size;
+                      row[column_index].char_data =
+                      (SQLCHAR *)
+                      realloc
+                      (
+                        row[column_index].char_data,
+                        row[column_index].size + data_to_copy_length
+                      );
+                      memcpy
+                      (
+                        row[column_index].char_data + row[column_index].size,
+                        data->long_data_buffer,
+                        data_to_copy_length
+                      );
+                    case SQL_C_WCHAR:
+                      data_to_copy_length = strlen16((const char16_t *)data->long_data_buffer) * 2;
+                      row[column_index].wchar_data =
+                      (SQLWCHAR *)
+                      realloc
+                      (
+                        row[column_index].wchar_data,
+                        row[column_index].size + data_to_copy_length
```

Comment:
Heads up! If `realloc` fails, it returns NULL *but does not free the original pointer*. As you have coded it, the original memory will be lost and leaked.

Reply:
This is a tricky one... a failure on a `realloc` seems that it should return a failure from this function (`fetch_and_store`). Of course returning `SQL_ERROR` will probably be misinterpreted as an error originating from the ODBC layer. Which normally wouldn't be a problem, but if an error was returned from this function when the `QueryAsyncWorker` is calling a query, it will throw a generic error to the user about an error retrieving result set:

https://github.com/markdirish/node-odbc/blob/bd8bb696232bd1e8a24e8626b6152bf9462b59e3/src/odbc_connection.cpp#L914-L918

Seems that maybe I need to create a special return code for use with memory allocation issues. The driver of course has `HY001` to throw when memory allocation errors happen, but not sure the best practice when memory allocation issues happen in the application layer.

Reply:
Have a method of checking allocs/reallocs now, and also realloc to a temp variable so that I don't lose the original pointer that will need to be freed.